src/NHibernate/Proxy/DynamicProxy/ProxyCacheEntry.cs

```diff
@@ -43,15 +44,26 @@ public ProxyCacheEntry(System.Type baseType, System.Type[] interfaces)
 		public System.Type BaseType { get; private set; }
 		public System.Type[] Interfaces { get; private set; }
 
+		[NonSerialized]
+		private HashSet<System.Type> _uniqueInterfaces;
+
 		public override bool Equals(object obj)
 		{
 			var that = obj as ProxyCacheEntry;
-			if (ReferenceEquals(null, that))
+			if (ReferenceEquals(null, that) ||
+				hashCode != that.hashCode ||
+				BaseType != that.BaseType)
 			{
 				return false;
 			}
 
-			return hashCode == that.GetHashCode();
+			// In case the instance was de-serialized, the _uniqueInterfaces will be null. Re-fetch its _uniqueInterfaces if needed.
+			if (_uniqueInterfaces == null)
+				_uniqueInterfaces = new HashSet<System.Type>(Interfaces);
+			if (that._uniqueInterfaces == null)
+				that._uniqueInterfaces = new HashSet<System.Type>(that.Interfaces);
+
+			return _uniqueInterfaces.SetEquals(that._uniqueInterfaces);
```

Comment:
`SetEquals` does not require the argument to be `ISet<>`, so the second if (`if (that._uniqueInterfaces == null)`) is not required